expect/_custom_equality_tester_test.ts

```diff
@@ -0,0 +1,84 @@
+// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
+
+import { expect } from "./expect.ts";
+
+class Duration {
+  public time: number;
+  public unit: "H" | "M" | "S";
+
+  constructor(time: number, unit: "H" | "M" | "S") {
+    this.time = time;
+    this.unit = unit;
+  }
+
+  toString(): string {
+    return `[Duration: ${this.time.toString()}${this.unit}]`;
+  }
+
+  equals(other: Duration): boolean {
+    if (this.unit === other.unit) {
+      return this.time === other.time;
+    }
+
+    if (
+      (this.unit === "H" && other.unit === "M") ||
+      (this.unit === "M" && other.unit === "S")
+    ) {
+      return (this.time * 60) === other.time;
+    }
+
+    if (
+      (other.unit === "H" && this.unit === "M") ||
+      (other.unit === "M" && this.unit === "S")
+    ) {
+      return (other.time * 60) === this.time;
+    }
+
+    return (this.time * 60 * 60) === other.time;
+  }
+}
+
+function isDurationMatch(a: Duration, b: Duration) {
+  const isDurationA = a instanceof Duration;
+  const isDurationB = b instanceof Duration;
+
+  if (isDurationA && isDurationB) return a.equals(b);
+  if (isDurationA === isDurationB) return undefined;
+  return false;
+}
+
+const duration1 = new Duration(1, "S");
+const duration2 = new Duration(2, "M");
+const duration3 = new Duration(120, "S");
+
+expect.addEqualityTester([isDurationMatch]);
+
+function* toIterator<T>(array: Array<T>): Iterator<T> {
+  for (const obj of array) {
+    yield obj;
+  }
+}
+
+Deno.test("Basic custom equality test", () => {
+  expect(duration1).toEqual(duration1);
+  expect(duration1).not.toEqual(duration3);
+  expect(duration1).not.toEqual(duration2);
+  expect(duration2).not.toBe(duration1);
+  expect(toIterator([duration1, duration2])).toEqual(
+    toIterator([duration2, duration1]),
+  );
```

Comment:
If I tested this on jest, it threw with the below error:
```
TypeError: 'caller', 'callee', and 'arguments' properties may not be accessed on strict mode functions or the arguments objects for calls to them
```
What does this case try to test? Also the inverted order of `duration1` and `duration2` intentional?

Reply:
Oooh, sorry I carelessly ignored this case. 

This case should _**not**_ equal **actually**.

What the test wants to check includes:
* check yield value in array is pass or not (whatever the order);
* check type is pass or not;
* check the instance indeed value is pass or not;

`Jest` implements this through [iterableEquality](https://github.com/jestjs/jest/blob/main/packages/expect-utils/src/utils.ts#L173), and `vitest` reuses this methods. And currently I don't have many ideas about this, so I marked it `todo`.

Or we also can reused the `jest` implementations?

WDYT?

Reply:
Reusing iterableEquality sounds good to me.

BTW this PR looks enough useful at this point. I'm landing this as is.

PRs for the above addition are welcome

Reply:
PR for this case: https://github.com/denoland/deno_std/pull/4286